Documentation/Evolution/UnicodeForStringProcessing.md

```diff
@@ -0,0 +1,677 @@
+
+Operating over classes of characters is a vital component of string processing. Swift's `String` provides, by default, a view of `Character`s or [extended grapheme clusters][graphemes] whose comparison honors [Unicode canonical equivalence][canoneq].
+
+```swift
+let str = "Cafe\u{301}" // "Café"
+str == "Café"           // true
+str.dropLast()          // "Caf"
+str.last == "é"         // true (precomposed e with acute accent)
+str.last == "e\u{301}"  // true (e followed by composing acute accent)
+```
+
+At a regular expression's simplest, without metacharacters or special features, matching should behave like a test for equality. A string should always match a regular expression that simply contains the same characters.
+
+```swift
+str.contains(/Café/)        // true
+```
+
+And from there, small changes should continue to comport with the element counting and comparison expectations set by `String`:
+
+```swift
+str.contains(/Caf./)        // true
+str.contains(/.+é/)         // true
+str.contains(/.+e\u{301}/)  // true
+str.contains(/\w+é/)        // true
+```
+
+With these initial principles in hand, we can look at how character classes should behave with Swift strings. Unicode leaves all interpretation of grapheme clusters up to implementations, which means that Swift needs to define any semantics for its own usage. Since other regex engines operate, at most, at the semantics level of Unicode scalar values, there is little to no prior art to consult.
+
+<details><summary>Other engines</summary>
+
+Character classes in other languages match at either the Unicode scalar value level, or even the code unit level, instead of recognizing grapheme clusters as characters. When matching the `.` character class, other languages will only match the first part of an `"e\u{301}"` grapheme cluster. Some languages, like Perl, Ruby, and Java, support an additional `\X` metacharacter, which explicitly represents a single grapheme cluster.
+
+| Matching  `"Cafe\u{301}"` | Pattern: `^Caf.` | Remaining | Pattern:  `^Caf\X` | Remaining |
+|---|---|---|---|---|
+| C#, Rust, Go | `"Cafe"` | `"´"` | n/a | n/a |
+| NSString, Java, Ruby, Perl | `"Cafe"` | `"´"` | `"Café"` | `""` |
+
+Other than Java's `CANON_EQ` option, the vast majority of other languages and engines are not capable of comparing with canonical equivalence.
+
+</details>
+
+## Proposed solution
+
+TK: semantic levels, options for controlling, canonical equivalence, Unicode properties
+
+## Detailed design
+
+First, we'll discuss the options that let you control a regex's behavior, and then explore the character classes that define the your pattern.
+
+### Options
+
+Options can be declared in two different ways: as part of [regular expression literal syntax][literals], or applied as methods when declaring a `Regex`. For example, both of these `Regex`es are declared with case insensitivity:
```

Comment:
I think Michael noted this on Slack already, but they're not part of the _literal_ syntax, they're part of the interior syntax (i.e. they work with string inits too).